src/Microsoft.AspNet.Mvc.Core/Rendering/Html/DefaultHtmlGenerator.cs

```diff
@@ -123,7 +123,7 @@ public virtual TagBuilder GenerateAntiForgery([NotNull] ViewContext viewContext)
                 metadata,
                 name,
                 value: "true",
-                useViewData: !explicitValue,
+                useViewData: (metadata == null && !explicitValue),
```

Comment:
Maybe add a comment when `metadata` could be null? Might make it easier to grock why that would cause `useViewData` to change.